### PR TITLE
Fix #16343 - Do not generate numeric structs from unnamed typedefs in "ts" ##types

### LINKED ISSUE #16343: unnamed struct named as 0 and no difference with struct and typedef
```
pmb:core pancake$ r2 -
 -- Buy a Mac
[0x00000000]> "td typedef struct {int a;} Foo;"
[0x00000000]> ts
0
Foo
[0x00000000]> q
pmb:core pancake$ r2 -
 -- There's more than one way to skin a cat
[0x00000000]> "td typedef struct Anon{int a;} Foo;"
[0x00000000]> ts
Anon
Foo
[0x00000000]>
```

### PR DESCRIPTION
**Pull request by oxcabe**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Numeric named structs shouldn't be appearing in "ts" commands.

**Test plan**

Create a test asserting that a numeric named struct generated after typedeffing anonymous struct does not show up in "ts" commands.

**Closing issues**

Closes https://github.com/radareorg/radare2/issues/16343.

### CODE REVIEW

**Comment by oxcabe** [comment]
Please check the output of "tsc" after the typedef. What should I do there?

**Comment by radare** [comment]
What about preventing this 0 to be stored in first? What happens if the key is removed? I think imwhatever you do is always a 0. So no need to check for isNumber() imho

**Comment by oxcabe** [comment]
I tried that earlier, it cannot find the type "Foo" if we get rid of all the information the parser generates about "0".

**Comment by oxcabe** [comment]
It's not always 0. Try to do:
```
"td typedef struct {int a;} Foo;"
"td typedef struct {int b;} Bar;"
"td typedef struct {int c;} Baz;"
```

**Comment by XVilka** [comment]
@radare it has to be created like that so the TCC will be able to reuse these types int he further definitions. It will be solved if/when we drop `shlr/tcc` and just migrate to tree-sitter for this parser as well.

**Comment by radare** [comment]
Please add a test, also notice the new typedef is redefined twice as int and can be accessed by int, so imho it should not be even saved by libtcc, i dont think tcc needs tha tinternal id at all

```
$ r2 -
 -- Select your character: RBin Wizard, Master Anal Paladin, or Assembly Warrior
[0x00000000]> "td typedef struct {int a;} Foo;"
[0x00000000]>
[0x00000000]> ts
0
Foo
NSString
[0x00000000]> ts 0
pf d a
[0x00000000]> ts Foo
pf d a
[0x00000000]> "td typedef struct {int a;} Foo;"
[0x00000000]> ts
0
1
Foo
NSString
[0x00000000]>
```

**Comment by radare** [comment]
Screenshot 2020-04-12 at 04 24 56

**Comment by radare** [comment]
see https://github.com/radareorg/radare2/pull/16539

**Comment by XVilka** [comment]
https://github.com/radareorg/radare2/pull/16485/checks?check_run_id=582172306#step:11:628
```
[XX] db/cmd/types typedef struct
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'to ../bins/headers/12272.h
ts 0
ts 1
?e ----
ts A
ts B
ts O  # Broken but shouldn't hang
?e ----
tk~^0,1
tk~typedef.A=,typedef.B=,typedef.O=,struct.0,struct.1,struct.A,struct.B,struct.O
?e
ts~0,1,A,B,O
' -
-- stdout
--- .a	2020-04-13 08:51:15.000000000 +0000
+++ .b	2020-04-13 08:51:15.000000000 +0000
@@ -1,22 +1,10 @@
-pf dF num num2
-pf c ch
----
-pf dF num num2
-pf c ch
+pf f num
+pf dFc num num2 ch
+pf dFc num num2 ch
----
-0=struct
-1=struct
-struct.0=num,num2
-struct.0.!size=96
-struct.0.num=int32_t,0,0
-struct.0.num.meta=0
-struct.0.num2=double,8,0
-struct.0.num2.meta=10
-struct.1=ch
-struct.1.!size=8
-struct.1.ch=char,0,0
-struct.1.ch.meta=2
struct.A=num
+struct.A.!size=32
struct.A=num
+struct.A.!size=32
struct.A.num=float,0,0
struct.A.num.meta=9
struct.O=ptr
@@ -26,8 +14,6 @@
typedef.B=struct 1
typedef.O=struct O

-0
-1
A
B
O


-- stderr
Cannot find '0' type
Cannot find '1' type

**Comment by oxcabe** [comment]
I'm on it

**Comment by radare** [comment]
pls rebase

File: shlr/tcc/tccgen.c [hunk 1]
```diff
@@ -3182,8 +3192,15 @@ static int decl0(int l, int is_for_loop_init) {
 					char buf[500];
 					alias = get_tok_str(v, NULL);
 					type_to_str(buf, sizeof(buf), &sym->type, NULL);
-					tcc_appendf("%s=typedef\n",alias);
-					tcc_appendf("typedef.%s=%s\n",alias, buf);
+					tcc_appendf ("%s=typedef\n", alias);
+					tcc_appendf ("typedef.%s=%s\n", alias, buf);
+					if (tcc_typedefs) {
+						while (*tcc_typedefs != NULL) { // XXX it should not enter here
+							tcc_appendf (*tcc_typedefs, alias);
```
Comment (oxcabe) [inline shlr/tcc/tccgen.c:3199]: segfaults here

Comment (oxcabe) [reply shlr/tcc/tccgen.c:3199]: @radare I don't really understand why this happens. Isn't the condition above checking if the pointer holds valid data? should I use calloc to allocate tcc_typedefs?

Comment (radare) [reply shlr/tcc/tccgen.c:3199]: the != NULL is unnecessary just `while (*tcc_typedefs)`

File: shlr/tcc/tccgen.c [hunk 2]
```diff
@@ -1178,13 +1182,19 @@ static void struct_decl(CType *type, int u) {
...
-							tcc_appendf ("[+]%s.%s=%s\n",
-								ctype, name, varstr);
-							tcc_appendf ("%s.%s.%s.meta=%d\n",
-								ctype, name, varstr, type_bt);
-							/* compact form */
-							tcc_appendf ("%s.%s.%s=%s,%d,%d\n",
-								ctype, name, varstr, b, offset, arraysize);
+							if (is_typedef && autonamed) {
+								tcc_typedef_appendf ("[+]typedef.%%1$s.fields=%s\n", varstr);
+								tcc_typedef_appendf ("typedef.%%1$s.%s.meta=%d\n", varstr, type_bt);
```
Comment (radare) [inline shlr/tcc/tccgen.c:1187]: why %%1$s ? wtf

File: shlr/tcc/tccgen.c [hunk 3]
```diff
@@ -3182,8 +3192,15 @@ static int decl0(int l, int is_for_loop_init) {
...
 					alias = get_tok_str(v, NULL);
 					type_to_str(buf, sizeof(buf), &sym->type, NULL);
-					tcc_appendf("%s=typedef\n",alias);
-					tcc_appendf("typedef.%s=%s\n",alias, buf);
+					tcc_appendf ("%s=typedef\n", alias);
+					tcc_appendf ("typedef.%s=%s\n", alias, buf);
+					if (tcc_typedefs) {
+						while (*tcc_typedefs != NULL) { // XXX it should not enter here
+							tcc_appendf (*tcc_typedefs, alias);
+							free (*tcc_typedefs);
```
Comment (radare) [inline shlr/tcc/tccgen.c:3200]: uhm you are freeing it after appending it, this sounds like a problem

File: libr/util/utype.c [hunk 4]
```diff
@@ -389,89 +470,14 @@ R_API char *r_type_format(Sdb *TDB, const char *t) {
...
+		return fmt_struct_union(TDB, var, false);
 	}
 	if (!strcmp (kind, "typedef")) {
 		snprintf (var2, sizeof (var2), "typedef.%s", t);
 		const char *type = sdb_const_get (TDB, var2, NULL);
 		// only supports struct atm
-		if (type && r_str_startswith (type, "struct ") && strcmp (type + 7, t)) {
-			return r_type_format (TDB, type + 7);
+		if (type && r_str_startswith (type, "struct ")) {
+			return fmt_struct_union(TDB, var, true);
```
Comment (radare) [inline libr/util/utype.c:480]: missing space before (

File: libr/util/utype.c [hunk 5]
```diff
@@ -372,11 +372,92 @@ static void filter_type(char *t) {
...
+	char *nfields = (is_typedef) ? fields : var;
+	for (n = 0; (p = sdb_array_get (TDB, nfields, n, NULL)); n++) {
+		char *type;
+		char *struct_name;
+		const char *tfmt = NULL;
+		bool isStruct = false;
+		bool isEnum = false;
+		bool isfp = false;
+		snprintf (var2, sizeof (var2), "%s.%s", var, p);
+		type = sdb_array_get (TDB, var2, 0, NULL);
```
Comment (radare) [inline libr/util/utype.c:389]: assign and define in noe line. char *type = sdb_...

File: shlr/tcc/tccgen.c (hunk 2 above)
Comment (oxcabe) [reply shlr/tcc/tccgen.c:1187]: I'ts a bit wild, but it works. I put it this way because these fields need the typedef identifier later on.

**Comment by radare** [comment]
still red. please rebase this pr and make it work

File: libr/util/utype.c [hunk 6]
```diff
@@ -372,11 +372,91 @@ static void filter_type(char *t) {
...
+				isfp = true;
+			} else if (!strncmp (type, "struct ", 7)) {
+				struct_name = type + 7;
+				// TODO: iterate over all the struct fields, and format the format and vars
+				snprintf (var3, sizeof (var3), "struct.%s", struct_name);
+				tfmt = sdb_const_get (TDB, var3, NULL);
+				isStruct = true;
+			} else {
+				// special case for char[]. Use char* format type without *
+				if (!strncmp (type, "char", 5) && elements > 0) {
```
Comment (radare) [inline libr/util/utype.c:405]: is this r_str_startswith() or strcmp ? because 5 is checking for the null byte

File: libr/util/utype.c [hunk 7]
```diff
@@ -372,11 +372,91 @@ static void filter_type(char *t) {
...
+					tfmt = sdb_const_get (TDB, "type.char *", NULL);
+					if (tfmt && *tfmt == '*') {
+						tfmt++;
+					}
+				} else {
+					if (!strncmp (type, "enum ", 5)) {
```
Comment (radare) [inline libr/util/utype.c:411]: use r_str_startswith

File: libr/util/utype.c [hunk 8]
```diff
@@ -372,11 +372,91 @@ static void filter_type(char *t) {
...
+					vars = r_str_append (vars, " ");
+				} else {
+					fmt = r_str_append (fmt, tfmt);
+					vars = r_str_append (vars, p);
+					vars = r_str_append (vars, " ");
+				}
+			} else {
+				eprintf ("Cannot resolve type '%s'\n", var3);
+			}
+		free (type);
```
Comment (radare) [inline libr/util/utype.c:447]: wrong indent

File: shlr/tcc/libtcc.c [hunk 9]
```diff
@@ -801,3 +801,23 @@ PUB_FUNC void tcc_appendf(const char *fmt, ...) {
 	tcc_cb (b, tcc_cb_ptr);
 	va_end (ap);
 }
+
+PUB_FUNC void tcc_typedef_appendf(const char *fmt, ...) {
+	char **typedefs_start;
+	if (!tcc_typedefs) {
+		tcc_typedefs = (char **)calloc (50, 1024);
```
[the remark this answers is not in the document]
Comment (oxcabe) [reply shlr/tcc/libtcc.c:808]: What should I do? Can I use RList for that?

**Comment by radare** [comment]
We can check for bondaries, but RPVector is probably a better option in this case

File: libr/util/utype.c [hunk 10]
```diff
@@ -372,11 +372,90 @@ static void filter_type(char *t) {
 	}
 }
 
+static char *fmt_struct_union(Sdb *TDB, char *var, bool is_typedef) {
+	// assumes var list is sorted by offset.. should do more checks here
+	char *p = NULL, *vars = NULL, var2[132], *fmt = NULL;
```
Comment (lowlyw) [inline libr/util/utype.c:377]: what's with the magic 132 size??

File: libr/util/utype.c [hunk 11]
```diff
@@ -372,11 +372,90 @@ static void filter_type(char *t) {
...
+		const char *tfmt = NULL;
+		bool isStruct = false;
+		bool isEnum = false;
+		bool isfp = false;
+		snprintf (var2, sizeof (var2), "%s.%s", var, p);
+		size_t alen = sdb_array_size (TDB, var2);
+		int elements = sdb_array_get_num (TDB, var2, alen - 1, NULL);
+		char *type = sdb_array_get (TDB, var2, 0, NULL);
+		if (type) {
+			char var3[128] = {0};
```
Comment (lowlyw) [inline libr/util/utype.c:392]: again, why magic size?

File: libr/util/utype.c [hunk 12]
```diff
@@ -372,11 +372,90 @@ static void filter_type(char *t) {
...
+				if (elements > 0) {
+					fmt = r_str_appendf (fmt, "[%d]", elements);
+				}
+				if (isStruct) {
+					fmt = r_str_append (fmt, "?");
+					vars = r_str_appendf (vars, "(%s)%s", struct_name, p);
+					vars = r_str_append (vars, " ");
+				} else if (isEnum) {
+					fmt = r_str_append (fmt, "E");
+					vars = r_str_appendf (vars, "(%s)%s", type + 5, p);
```
Comment (lowlyw) [inline libr/util/utype.c:436]: get rid of magic 5, use sizeof () for a constant for the "enum " array.

File: libr/util/utype.c [hunk 13]
```diff
@@ -372,11 +372,90 @@ static void filter_type(char *t) {
...
+			} else {
+				// special case for char[]. Use char* format type without *
+				if (!strcmp (type, "char") && elements > 0) {
+					tfmt = sdb_const_get (TDB, "type.char *", NULL);
+					if (tfmt && *tfmt == '*') {
+						tfmt++;
+					}
+				} else {
+					if (r_str_startswith (type, "enum ")) {
+						snprintf (var3, sizeof (var3), "%s", type + 5);
```
Comment (lowlyw) [inline libr/util/utype.c:411]: get rid of magic 5, use a constant for the size of `"enum "`

File: libr/util/utype.c [hunk 14]
```diff
@@ -372,11 +372,90 @@ static void filter_type(char *t) {
...
+	free (vars);
+	return fmt;
+}
+
 R_API char *r_type_format(Sdb *TDB, const char *t) {
-	int n;
-	char *p, var[130], var2[132];
-	char *fmt = NULL;
-	char *vars = NULL;
+	char var[130], var2[132];
```
Comment (lowlyw) [inline libr/util/utype.c:458]: can you `vasprintf` and free? i'm not sure if this is *needed* but it's kind of weird to have these magic buffer sizes.

File: shlr/tcc/tccgen.c [hunk 15]
```diff
@@ -755,17 +756,16 @@ static void type_to_str(char *buf, int buf_size, CType *type, const char *varstr
...
 		if (bt == VT_STRUCT) {
-			tstr = "struct ";
+			tstr = "struct";
 		} else if (bt == VT_UNION) {
-			tstr = "union ";
+			tstr = "union";
 		} else {
-			tstr = "enum ";
+			tstr = "enum";
 		}
```
Comment (lowlyw) [inline shlr/tcc/tccgen.c:764]: just use the fallthrough here, and set tstr on each of these, it's weirdly done in this case with the if when you already know the type.

like 

```
case VT_ENUM:
  tstr = "enum";
  // allow fallthrough
case VT_STRUCT:
  tstr = "struct";
  // allow fallthrough
```

also maybe there should be constants for these strings?

File: libr/util/utype.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (oxcabe) [reply libr/util/utype.c:411]: Probably not, that `type` variable holds a SDB value related to types generated by TCC

File: shlr/tcc/libtcc.c [hunk 16]
```diff
@@ -801,3 +802,27 @@ PUB_FUNC void tcc_appendf(const char *fmt, ...) {
...
 }
+
+PUB_FUNC void tcc_typedef_appendf(const char *fmt, ...) {
+	if (!tcc_typedefs) {
+		tcc_typedefs = r_pvector_new ((RPVectorFree) free);
+	}
+	char typedefs_tail[1024];
+	va_list ap;
+	va_start (ap, fmt);
+	if (vsnprintf (typedefs_tail, sizeof (typedefs_tail), fmt, ap) > 0) {
```
[the remark this answers is not in the document]
Comment (oxcabe) [reply shlr/tcc/libtcc.c:813]: Fixed !

[the remark this answers is not in the document]
Comment (oxcabe) [reply shlr/tcc/libtcc.c:813]: btw `vasprintf` is neat, I didn't know about it

**Comment by oxcabe** [comment]
Regarding `asprintf` function family (from Linux Programmer's Manual): 
> These functions are GNU extensions, not in C or POSIX.  They are also available under *BSD.

I asked @radare if we can implement our own POSIX compliant API functions, for example `r_str_vasprintf(char **strp, const char *fmt, va_list ap)`.

**Comment by ret2libc** [comment]
You can use `r_str_newf` instead of `vasnprintf`.

**Comment by radare** [comment]
No, he cant, because tcc cant link with rutil. Is util who links to tcc

**Comment by ret2libc** [comment]
> No, he cant, because tcc cant link with rutil. Is util who links to tcc

Right. In tcc you can't, but in libr/util/utype.c you can if you want.

**Comment by radare** [comment]
Imho the pr is fine to merge it like this,  i dont expect anyone to be defining structs with 4K chars in length, so a reasonable string length restriction is fine for me.

About adding vaasprintf in r2 ive never felt the need to use it, but if added i would call it r_str_vnewf() and this (in case of done) which i dont think we need it. Must be done in a separate PR with proper tests and such

**Pull request merged**